src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/UserDefinedTypeDescriptor.cs

```diff
@@ -41,6 +41,110 @@ public uint GetVariableTypeIndex(TypeDesc type)
             }
         }
 
+        public uint GetStateMachineThisVariableTypeIndex(TypeDesc type)
+        {
+            type = DebuggerCanonicalize(type);
+
+            lock (_lock)
+            {
+                if (!_knownStateMachineThisTypes.TryGetValue(type, out uint typeIndex))
+                {
+                    Debug.Assert(type.IsDefType);
+
+                    MetadataType defType = (MetadataType)type;
+                    ClassTypeDescriptor classTypeDescriptor = new ClassTypeDescriptor
+                    {
+                        IsStruct = 1,
+                        //IsStruct = type.IsValueType ? 1 : 0,
+                        Name = $"StateMachineLocals_{System.Reflection.Metadata.Ecma335.MetadataTokens.GetToken(((EcmaType)defType.GetTypeDefinition()).Handle):X}",
+                        //BaseClassId = type.HasBaseType && !type.IsValueType ? GetTypeIndex(defType.BaseType, true) : 0,
+                        InstanceSize = defType.InstanceByteCount.IsIndeterminate ? 0 : (ulong)defType.InstanceByteCount.AsInt,
+                    };
+
+                    var fieldsDescs = new ArrayBuilder<DataFieldDescriptor>();
+
+                    foreach (var fieldDesc in defType.GetFields())
+                    {
+                        if (fieldDesc.IsStatic)
+                            continue;
+
+                        string fieldNameEmit = fieldDesc.Name;
+                        if (fieldNameEmit.Length > 0 && fieldNameEmit[0] == '<')
+                        {
+                            if (TryGetGeneratedNameKind(fieldNameEmit, out char kind))
```

Comment:
I assume that this logic is copied from somewhere, or at least inspired by some code in Roslyn (?). Add a comment with a link for where this came from?

Reply:
Ah yes. I had such comment but it looks like it fell victim to when I was updating this. Fixed.